src/main/java/org/jboss/pnc/cleaner/archiver/BuildArchiver.java

```diff
@@ -0,0 +1,437 @@
+import org.jboss.pnc.dto.ProductMilestone;
+import org.jboss.pnc.dto.ProductMilestoneRef;
+import org.jboss.pnc.dto.ProductRef;
+import org.jboss.pnc.dto.ProductVersion;
+import org.jboss.pnc.enums.BuildStatus;
+import org.jboss.pnc.rest.api.parameters.BuildsFilterParameters;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+import javax.transaction.Transactional;
+import javax.ws.rs.core.Response;
+import java.io.BufferedReader;
+import java.time.temporal.ChronoField;
+import java.time.temporal.IsoFields;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+import static org.jboss.pnc.cleaner.archiver.ArchviedBuildRecord.ErrorGroup.INDY;
+import static org.jboss.pnc.cleaner.archiver.ArchviedBuildRecord.ErrorGroup.ND;
+import static org.jboss.pnc.cleaner.archiver.ArchviedBuildRecord.ErrorGroup.PNC;
+import static org.jboss.pnc.cleaner.archiver.ArchviedBuildRecord.ErrorGroup.PSI;
+import static org.jboss.pnc.cleaner.archiver.BuildArchiverStrings.*;
+
+/**
+ * @author <a href="mailto:matejonnet@gmail.com">Matej Lazar</a>
+ */
+@ApplicationScoped
+public class BuildArchiver {
+    private final Logger logger = LoggerFactory.getLogger(BuildArchiver.class);
+    public static final String BUILD_ARCHIVED = "BUILD_ARCHIVED";
+    @Inject
+    @RestClient
+    FinalLogClient finalLogs;
+    @Inject
+    BuildClient buildClient;
+    @Inject
+    ProductMilestoneClient productMilestoneClient;
+    @Inject
+    ProductVersionClient productVersionClient;
+    @Inject
+    BuildArchiver self;
+    @ConfigProperty(name = "buildArchiver.trimmedLogMaxSize", defaultValue = "1000000")
+    private Integer trimmedLogMaxSize;
+
+    @Timed
+    @Scheduled(cron = "{buildArchiverScheduler.cron}")
+    public void archiveBuilds() {
+        logger.info("Archiving Builds ...");
+        Collection<Build> unverifiedBuilds = getUnarchivedBuilds().getAll();
+        logger.info("Found {} unverified builds.", unverifiedBuilds.size());
+        unverifiedBuilds.forEach(build -> archiveBuild(build));
+    }
+
+    @Timed
+    private RemoteCollection<Build> getUnarchivedBuilds() {
+        BuildsFilterParameters buildsFilterParameters = new BuildsFilterParameters();
+        buildsFilterParameters.setRunning(false);
+        List<String> attributes = Collections.singletonList("!" + BUILD_ARCHIVED);
+        try {
+            return buildClient.getAll(buildsFilterParameters, attributes);
+        } catch (RemoteResourceException e) {
+            logger.error("Cannot read remote builds.", e);
+            return RemoteCollection.empty();
+        }
+    }
+
+    @Timed
+    public void archiveBuild(Build build) {
+        long trimLogSize = 0;
+        if (build.getTemporaryBuild() && build.getStatus() != BuildStatus.SUCCESS) {
+            trimLogSize = trimmedLogMaxSize;
+        }
+
+        logger.info("Reading alignment logs");
```

Comment:
1) If I am reading this correctly, the download of the alignment and build logs, and their categorization, is done for every build. I think this can be improved to match the logic of the current trigger, which analyzes the log ONLY for the builds which have a SYSTEM_ERROR or a FAILED status. This will reduce a lot the number of processed logs and speed up the archival (all builds will still be archived, but only a subset will have their logs analyzed).

The category for the builds which are not analyzed (so all builds not having status SYSTEM_ERROR or FAILED) is left empty.

2) The `trimmed_alignment_log` is actually not trimmed at all in the trigger because it might not contain `Caught exception:` part

Reply:
1) we need to read the logs always, because:
  - build logs: we want to detect if it is `MVN-WRAPPED-NPM`
  - alignment logs: we want to detect manipulation en/disabled and brew pull state
2) I know, the trimming here works on 2 levels, if it detects `Caught exception:` then it throw away anything before that and it always limits the total size to the last X characters (roughly bytes, unles there is a lot of nonASCII chars). X is configurable, defaults to 1M. I think I heard about cases with extremly large repour logs, so I decided to limit the size to be safe.

Reply:
You are absolutely correct about 1, I forgot about the very initial part of the trigger. All good then!